### DISCUSSION TITLE
Thoughts on new DAG Grid

### DISCUSSION THREAD

**Discussion by tagoad** [apache/airflow#23413] [Ideas] [answered]
Just updated to 2.3 and was surprised by the replacement of the tree view with the grid view, and i wanted to get some conversation going about pros and cons of it. While i appreciate the cleanness of it, I do have some issues with it:

- Task Relationships: This is my major issue with it, there isn't anything to show the relationship between tasks. Maybe some grouping based on task groups could be added? Alternates colors similar to banded table rows to show difference in task groups?
- The run information is right aligned which leaves a ton of whitespace between the task names and their run info. Literally have my task names far left and then the info far right, not great for ease of reading. 

I do really like the details panel, and I think that opens up a lot of possibilities for at a glance info for individual task you select.

**Comment by tagoad** [comment]
Looking at the new docs i do see that there is some functionality with Task Groups and the grid view so i am going to looking into that.

**Comment by bbovenzi** [comment]
Thanks for the feedback. Tree -> Grid was a big change, but I think it opens up interesting possibilities. 

<img width="396" alt="Screen Shot 2022-05-02 at 8 23 58 PM" src="https://user-images.githubusercontent.com/4600967/166346210-74c88cae-888c-4275-8155-26007189c1e5.png">

- Task Groups: You can expand/collapse them just like in the Graph view:
- There is also a "crosshair" highlight when you hover over a box to see what task and run it belongs to. But maybe we can make that more clear.
- Task relationships: Beyond groups and the order tasks are displayed, we don't show too much of the relationship. This is something I want to improve on. One idea is to add the Graph into the details panel.

**Reply by potiuk** [reply]
Yeah. I think this is a huge improvement and incrementally adding things to make the relationships more visible is cool. 

I think it would be great to see what is "really" missing there. That would likely help @bbovenzi and the team to approach it in most efficient way:

- is this the whole dependency graph visible in full while you are looking at task
- or maybe seeing dependencies while you hover/focus on specific task is enough ?

I also think there is a huge benefit of the Grid View that will become actually useful in the future - when we get into versioning. I think for TreeView it was next to impossible even to think that you could visualise that, while in Grid View - it's much easier to imagine :).

For me it looks like whatever we come up should be somewhat transient - i.e. the graph we should see should reflect which run and possibly even which task we are focused at a given time. The details view or hover are both possible to show it.

**Reply by argibbs** [reply]
Ooh, but so many thoughts. This is a real marmite of a change... 

My initial reaction, and one that's not gone away after a solid hour of use is "ug, I can't _do_ anything anymore". 😄 

The biggest problem for me is (as alluded to above) the lack of relationship in tasks. I note the grid appears to use the task groups to group by. The problem is that we use task groups heavily. When I open a graph view, I've always got something like:
![image](https://user-images.githubusercontent.com/4406514/166461553-afc70ed7-641a-4a73-8c16-c7176783c168.png)

Tree view used to let me x-ray this, but now with grid all I get is:
![image](https://user-images.githubusercontent.com/4406514/166461647-c3f9d3b4-83ca-4b04-aa20-3741bb4ee096.png)

Several of these nodes have groups within groups within groups. If there's a failure it takes me many clicks to dig down. And a task within one group may jump off to a task in another group.

Several off-the-cuff "issues":
1. No visual differentiation between "group" rows and "task" rows
2. No expand-all / expand-children options
3. No visual link between a selected node and it's children. Perhaps you could make all the immediate children / all descendents of any depth "glow", like you do when selecting tasks in the graph view.

I'll add more thoughts as I and my (vocal) internal clients spend more time with it.

It's got potential, but it will take time to get used to it. And we'll probably wait for some ui refinements before rolling beyond our dev env. It would be easier for us to migrate to grid if tree had been left in place, with grid added as well, not instead of it...

**Reply by potiuk** [reply]
I think this is fantastic feedback. Keep on getting more :). I think we are going to to a "quicker" turnaround on some changes in 2.3.0 (I've heard 2.3.1 in a week or so :) ) and maybe some of those comments **might** be incorporated sooner rather than later (but I am not the web person so not sure if/what would be possible ;) ). I don't want to promise anything as it's not me doing it :P

**Reply by tagoad** [reply]
Now that i have gotten our instance setup with TaskGroups i am liking the grid view more. I do agree with argibbs that there are some UI improvements that could be made to help differentiate between tasks, groups, and tasks within groups. A preview in the details pane of the relationships for a specific task would be really nice. Overall I do like it.

**Reply by ashb** [reply]
@argibbs Could you share a screenshot of how the Tree view looked for your dag for comparison? 

Cos one of the problems with the tree view showing dependencies was that it very quickly collapsed into a) a mess (with the same task in multiple places) or b) an incredibly deep tree

**Reply by ashb** [reply]
> A preview in the details pane of the relationships for a specific task would be really nice. Overall I do like it.

We have thought about bringing some (all?) of the Graph view into the new grid view -- we will explore ways of doing this so when you click on a Task/TaskInstance of showing something useful. (Right now I can't think of an easy way of showing it without a click without the drawbacks of the tree view)

**Reply by argibbs** [reply]
> @argibbs Could you share a screenshot of how the Tree view looked for your dag for comparison?
> 
Absolutely. Here you go (very slightly redacted, sorry). This is the same dag as before, only without the red for the latest dag run, as that was caused by #23411 and I've subsequently rolled back to get it to go green.

![image](https://user-images.githubusercontent.com/4406514/166570780-412298f1-7c45-47c6-80b3-a24c0d2ecc23.png)

**Reply by argibbs** [reply]
I totally agree with @ashb - the old tree rapidly descends into a mish-mash of duplicated nodes. We have some dags that wear the scroll wheel out they're that long. However, it does provide useful info if you are used to it. 

My personal opinion on what the grid "needs" is:
1. One-click fully expanded (ideally can make this by default).
2. Some indication of the node relationships. For me, I mostly care because I want to know if I clear the node what's going to be impacted. I _think_ it would be enough to "glow" all immediate children, but I'd need to use it to be sure I don't need all descendants of any depth, sorry.

--- 
Edit: thinking on it some more, yes, what I want to see is the immediate children. My use case is usually (as you can see from the failed dag runs above) "quickly find the failed nodes, assess how bad the rot is, clear the fails and retry". I do make heavy use of the future/past/upstream/failed options when clearing. Tree view does not scale and turns into a mélange of node, but what it _does_ almost always do is 'given a node, show you the immediate children' which helps immensely with the above task. Of course, you could do this by simply showing them in the info pane on the right, rather than trying to do highlights.

**Reply by potiuk** [reply]
> Edit: thinking on it some more, yes, what I want to see is the immediate children.

That was also my intuition - that users need to see the "immediate" children and possibly "immediate parents" of  task they are focusing on at any given moment, rather than full graph. 

And taking into account that in the future we will have versioning likely, I'd say this relation should be property of particular task execution not the whole task. I.e. - if you click (or hover over) particular execution or particular tasks, then you should. see (and be able to quickly navigate to) immediate parents and children in tree-like view for this particular task execution. 

Now. The difficul part is how to do it fast and intuitive :)

**Comment by KevinYang21** [comment]
Bringing the discussion in [this PR](https://github.com/apache/airflow/pull/18675) here:

The thing missing for me is the task depedency (I can collect more use cases from our internal users). Not the topology for the entire DAG, but a couple layers of the upstream/downstream for a couple of tasks. For those upstream/downstream tasks, what I do mostly are check the task status, check task log and clear task.

One example of my operation is:
- want to check the progress of my DAG
- go into tree view and see if the root task I care the most finished or not
- found it has not started and start to panic
- quickly eyeball the next a couple of rows (internally we view the depedency from downstream->upstream) to figure out which upstream tasks are still running
- found the upstream task stuck in running state because we ran out of EMR resources
- rush to our EMR team and scream for more resources

Edit 0:
Adding a quick use case from an internal client: they own a big DAG with a couple thoudsands of tasks, they don't use task group and thus use `root=task_id` to view only the task and it's upstreams. Maybe at least partial view with a root task can be done in grid view?

**Reply by argibbs** [reply]
> found it has not started and start to panic

LMAO. It me.

**Reply by ashb** [reply]
> Adding a quick use case from an internal client: they own a big DAG with a couple thoudsands of tasks, they don't use task group and thus use root=task_id to view only the task and it's upstreams. Maybe at least partial view with a root task can be done in grid view?

@KevinYang21 The Grid view respects the `?root=` parameter as before, and there's a "Filter Upstream" button.

**Reply by KevinYang21** [reply]
@ashb that's good to know, thanks!

**Comment by bbovenzi** [comment, marked as answer]
From all the discussion here are a few ideas I have:

1. Visually distinguish group summaries from their actual children tasks. I'm thinking something like below: (A previous version of the grid view used a background color but it got too noisy visually) PR: https://github.com/apache/airflow/pull/23488
<img width="507" alt="Screen Shot 2022-05-04 at 10 39 44 AM" src="https://user-images.githubusercontent.com/4600967/166708489-f36084ad-1df2-4947-b421-71179161c5cb.png">

2. We have a "Filter upstream" to see just the upstream of a task. Maybe we should have a filter downstream too. Also, we can include a mini-graph in the details panel to show immediate parents and children.
<img width="989" alt="Screen Shot 2022-05-04 at 10 55 46 AM" src="https://user-images.githubusercontent.com/4600967/166708889-d51db399-5df9-4085-bb00-2261c6a613f4.png">

3. If a DAG has task groups, add an expand all/collapse all button. PR: https://github.com/apache/airflow/pull/23487

4. We're adding more filters soon in https://github.com/apache/airflow/pull/23392. We can add filters for task instances that will also expand task groups to show that original failing task.

5. Once we update all the task action endpoints https://github.com/apache/airflow/issues/23412. Instead of just a text dump of what task instances are affected, we visually highlight the affected tasks in the Grid view.

**Reply by argibbs** [reply]
All the above sound great. 

Re:
> 1. ... A previous version of the grid view used a background color but it got too noisy visually

I'm not a gui guy, so I'll take your word for it. The line is a definite improvement.

**Reply by tagoad** [reply]
I love the bottom border idea, as well as the small graph in the details pane!

**Reply by bbovenzi** [reply]
Mini graph in the details panel (WIP):

![May-05-2022 13-49-08](https://user-images.githubusercontent.com/4600967/166985389-93637a60-c5b8-4941-9a6e-4c7d403f0488.gif)

**Reply by tagoad** [reply]
I wish I could give that more than one thumbs up. I do have a question though, specifically where it shows the upstream of the downstream task. We have a dummyoperator that cap all our dags, so for that there would be like 15+ upstream tasks branching from it and nothing below, how does it handle that?
Also, I noticed today that the Grid view is constantly reorganizing the tasks when auto-refresh is on. Is that normal?

**Reply by bbovenzi** [reply]
This WIP is just showing the whole dag but centering on the selected task.

The grid view reordering: that's probably a bug. Let me look into it.

**Reply by potiuk** [reply]
This is BEYOND COOL

**Reply by potiuk** [reply]
@KevinYang21 :) - WDYT ?

**Reply by argibbs** [reply]
Is the absolute size of the mini graph fixed? We have some biiiig graphs, if they shoved the task actions pane off the bottom I would cry.

**Reply by tagoad** [reply]
> This WIP is just showing the whole dag but centering on the selected task.
> 
> The grid view reordering: that's probably a bug. Let me look into it.

Do you want me to create an issue for it?

**Reply by bbovenzi** [reply]
> Is the absolute size of the mini graph fixed? We have some biiiig graphs, if they shoved the task actions pane off the bottom I would cry.

I want to keep the task actions in view. Right now, I'm exploring it with a fixed height and no zoom (but maybe a minimap). And adding a button to center on the task in question.

<img width="1263" alt="Screen Shot 2022-05-06 at 10 30 31 AM" src="https://user-images.githubusercontent.com/4600967/167156619-563e8808-1cde-4942-a329-bbb234b696d6.png">

**Reply by bbovenzi** [reply]
@tagoad yes, feel free to make an issue!

**Reply by potiuk** [reply]
Suggestion: I am not sure if it's easy/difficult to do, but for huge graphs there should be IMHO an option to just show direct upstream/downstream deps of selected task. in the case visible above you cannot directly see the upstream/downstream tasks and you loose a lot of the visualisation 'strength' this way.

**Comment by potiuk** [comment]
1. LGTM. Maybe another option is some icon indicating task group? But this is not strong. All the rest - cool.

**Reply by bbovenzi** [reply]
The up/down chevron is already that icon, no?

**Reply by potiuk** [reply]
right. the bottm border + bold is cool :)
